internal/worker/resizable_permit.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package worker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/marusama/semaphore/v2"
+)
+
+type resizablePermit struct {
+	sem semaphore.Semaphore
+}
+
+// NewResizablePermit creates a dynamic permit that's resizable
+func NewResizablePermit(initCount int) Permit {
+	return &resizablePermit{sem: semaphore.New(initCount)}
+}
+
+// Acquire is blocking until a permit is acquired or returns error after context is done
+// Remember to call Release(count) to release the permit after usage
+func (p *resizablePermit) Acquire(ctx context.Context) error {
+	if err := p.sem.Acquire(ctx, 1); err != nil {
+		return fmt.Errorf("failed to acquire permit before context is done: %w", err)
+	}
+	return nil
+}
+
+// Release release one permit
+func (p *resizablePermit) Release() {
+	p.sem.Release(1)
```

Comment:
how does semaphore behave if Release() is called multiple times? does it increase the capacity?

Reply:
It'll (essentially) increase capacity until it exceeds the limit, it's not a `sync.Once` or equivalent that can guarantee at-most-once-per-acquire.

When it exceeds the limit (possibly/probably at some later time), it panics: https://github.com/marusama/semaphore/blob/master/semaphore.go#L170

Reply:
there's a bit of a fundamental tradeoff between "ignore misuse, allow multiple calls" and "require exactly correct use, but it might be hard to find the cause".

tbh personally I prefer "require exactly correct use" in most cases.  it's easy to convert that to "ignore misuse" with a `sync.Once.Do(cancel)` wrapper, e.g. for stuff like https://github.com/cadence-workflow/cadence/blob/master/common/clock/ratelimiter.go#L354 where it's convenient to use both `defer` and early-unlock, and not have to worry about the many combinations possible.

Reply:
yea, there are a few unfortunately-complicated delayed-release things in the client :\  I'd love to get rid of as many as possible, or make them much clearer if not possible.